OrbitCore/CallstackData.cpp

```diff
@@ -32,5 +117,51 @@ void CallstackData::AddCallStackFromKnownCallstackData(
     absl::MutexLock lock(&unique_callstacks_mutex_);
     unique_callstacks_[hash] = std::move(unique_callstack);
   }
-  callstack_events_.push_back(CallstackEvent(event));
+  absl::MutexLock lock(&callstack_events_by_tid_mutex_);
```

Comment:
Might this lead to a deadlock?

Reply:
In what case? Can you elaborate?

Reply:
Haven't looked too closely, but you essentially have two locks in a sequence, which is always at least prune to deadlocks. @dimitry- maybe you might have a brief look.

Reply:
But the previous one has already been unlocked.

Reply:
I do not think there is a deadlock but do we want 2 separate mutexes in this class?

Reply:
Should I use a single mutex for the entire class?

Reply:
It most cases yes. The logic is goes like this (and there might be exceptions)

The data is the class is somehow interconnected so it should probably ensure that it does not end up in logically inconsistent state. And if data is not logically connected to each other - maybe it should be 2 classes instead of 1.

Reply:
There's a single mutex now, please have a look. Had to make it reentrant though (see the comment), so I'm using `std::recursive_mutex`.

Reply:
What's even the reason for not using reentrant locks by default? It seems to be unnecessarily prone to self deadlock.

Reply:
Ok, so I've read a lot of stuff against reentrant mutexes. Still, I think that given the `ForEach...` methods here, this is necessary. Do you have thoughts on this?